text/0000-min-rust-version.md

```diff
@@ -0,0 +1,205 @@
+field in the `Cargo.toml` sent to crates.io equals to "1.30" and for others to
+"1.40". Now if you'll build your project with e.g. Rust 1.33, `cargo` will select
+`foo v0.1.5`. `foo v0.1.9` will be selected only if you'll build your project with
+Rust 1.40 or higher. But if you'll try to build your project with Rust 1.29 cargo
+will issue an error.
+
+Note that described MSRV constraints and checks for dependency versions resolution
+can be disabled with `--no-msrv-check` option.
+
+`rust` field should respect the following minimal requirements:
+- value should be equal to "stable", "nightly: \*" or to a version in semver format.
+Note that "1.50" is a valid value and implies "1.50.0". (also see "nightly versions"
+extension)
+- version should not be bigger than the current stable toolchain
+- version should not be smaller than 1.27 (version in which  `package.rust` field
+became a warning instead of an error)
+
+`rust` will be a required field. For crates uploaded before introduction of this
+feature 2015 edition crates will imply `rust="1.0"` and 2018 edition will imply
+`rust = "1.30"`.
+
+It will be an error to use `rust="1.27"` and `edition="2018"`, but `rust="1.40"`
+and `edition="2015"` is a valid combination.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The describe functionality can be introduced in several stages:
+
+
+## First stage: dumb field
+
+At first the `rust` field can be simply a declarative optional field without any
+functionality behind it with minimal checks. The reason for it is to reduce
+implementation cost of the first stage to the minimum and ideally ship it as part
+of Rust 2018. It will also allow crate authors who care about MSRV to start mark
+their crates early.
+
+## Second stage: `cargo publish` check
+
+The next step is for `cargo publish` to require use of the toolchain specified
+in the `rust` field, for example crates with:
+- `rust="stable"` can be published only with a stable toolchain, though not
+necessarily with the latest one. Cargo will insert toolchain version before
+publishing the crate as was described in the "guide-level explanation".
+- `rust="nightly: *"` can be published only with a nightly toolchain. If finer
+grained "nightly: ..." (see "nightly versions" section) is selected, then one
+of the selected Nightly versions will have to be used.
+- `rust="1.30"` can be published only with (stable) Rust 1.30, even if it's
+not the latest stable Rust version.
+
+Using the usual build check `cargo publish` will verify that crate indeed can be
+built using specified MSRV. This check can be used with exisiting `--no-verify`
+option.
+
+## Third stage: versions resolution
+
+`rust` field will be used as a constraint for dependency versions resolution.
+If user uses e.g. Rust 1.40 and uses crate `foo = "0.2"`, but
+all selected versions of `foo` specify MSRV e.g. equal 1.41 or bigger (or even
+nightly) `cargo` will issue an error.
```

Comment:
What does this mean in this context:

If `foo` has published versions `0.2.1` (with `rust = 1.40`) and `0.2.2` (with `rust = 1.41`), and you're running rust `1.40`, depending on `foo = 0.2`, is the intention to:
- Error because you explicitly said to use the latest patch of `foo` is `0.2.2` and the latest patch requires a rust version greater than you are running.
- Build using `foo` `0.2.1` because it fits your wildcard and works for your rust version?

I would want to make sure your crate fails to build in this case because otherwise you might be using an insecure or error-prone version of a crate in this case even if the issue was patched.

Reply:
There should be a separate flag to indicate insecure versions, a bump in patch level is meaningless for determining if there are open security issues.  Security flaws can be discovered, but not fixed, so reporting error-prone versions should be separate from uploading new versions.

Reply:
If you write `foo = 0.2`, you're explicitly saying you want the latest patch version of `foo 0.2` available, even if that breaks your build.

Reply:
That's explicitly _not_ how it works.

* 0.2 is compatible with anything in the 0.2.z series
* If another crate anywhere in the build needs 0.2.1 and latest is 0.2.2, you get 0.2.1 even though it's not latest.

Reply:
If there is already precedent for not always pulling the latest, wouldn't rustc needing version X be akin to the create's needs?  Not that I need any specific behavior...

I join this thread when I discovered that I couldn't build RLS on Debian's rustc and the failure mode was to try to build and fail.  I'm most interested in the cases where rust does build something, but doesn't produce well and good binaries because it's lacking some feature.

Reply:
TIL it doesn't use the latest version if you don't specifically tell it to. My apologies!